mmengine/infer/infer.py

```diff
@@ -0,0 +1,468 @@
+        """Load config and weights from metafile.
+
+        Args:
+            model (str): model name defined in metafile.
+
+        Returns:
+            Tuple[Config, str]: Loaded Config and weights path defined in
+                metafile.
+        """
+        scope = DefaultScope.get_current_instance().scope_name  # type: ignore
+        assert scope is not None, ('scope should be initialized if you want '
+                                   'to load config from metafile.')
+        project = PKG2PROJECT[scope]
+        package_path = get_installed_path(project)
+        meta_indexes = load(osp.join(package_path, '.mim', 'model-index.yml'))
+        for meta_path in meta_indexes['Import']:
+            # meta_path example: mmcls/.mim/configs/conformer/metafile.yml
+            meta_path = osp.join(package_path, '.mim', meta_path)
+            metainfo = load(meta_path)
+            for model_cfg in metainfo['Models']:
+                if model_cfg['Name'] == model or model in model_cfg.get(
+                        'Alias', []):
+                    cfg = Config.fromfile(
+                        osp.join(package_path, '.mim', model_cfg['Config']))
+                    weights = model_cfg['Weights']
+                    return cfg, weights
+        raise ValueError(f'Cannot find model: {model} in {project}')
+
+    def _init_model(
+        self,
+        cfg: ConfigType,
+        weights: str,
+        device: str = 'cpu',
+    ) -> nn.Module:
+        """Initialize the model with the given config and checkpoint on the
+        specific device.
+
+        Args:
+            cfg (ConfigType): Config contained the model information.
+            weights (str): Path to the checkpoint.
+            device (str, optional): Device to run inference. Defaults to 'cpu'.
+
+        Returns:
+            nn.Module: Model loaded with checkpoint.
+        """
+        model = MODELS.build(cfg.model)
+        load_checkpoint(model, weights, map_location='cpu')
+        model.cfg = cfg.model
+        model.to(device)
+        model.eval()
+        return model
+
+    def _init_collate(self, cfg: ConfigType) -> Callable:
+        """Initialize the ``collate_fn`` with the given config.
+
+        The returned ``collate_fn`` will be used to collate the batch data.
+        If will be used in :meth:`preprocess` like this
+
+        .. code-block:: python
+            def preprocess(self, inputs, batch_size, **kwargs):
+                ...
+                dataloader = map(self.collate_fn, dataloader)
+                yield from dataloader
+
+        Args:
+            cfg (ConfigType): Config which could contained the `collate_fn`
+                information. If `collate_fn` is not defined in config, it will
+                be :func:`pseudo_collate`.
+
+        Returns:
+            Callable: Collate function.
+        """
+        try:
+            collate_fn = COLLATE_FUNCTIONS.get(cfg.test_dataloader.collate_fn)
```

Comment:
Does it support getting collate_fn from downstream repos?

Reply:
Changes to

```python
            with Registry.switch_scope_and_registry(self.scope) as registry:
                collate_fn = registry.get(
                    cfg.test_dataloader.collate_fn)
```